packages/worker/src/worker.ts

```diff
@@ -1644,21 +1671,43 @@ export class Worker {
    * @returns the result of `fnOrPromise`
    *
    * Throws on fatal Worker errors.
+   *
+   * **SDK versions `< 1.5.0`**:
+   * This method would not wait for worker to complete shutdown if the inner `fnOrPromise` threw an error.
+   *
+   * **SDK versions `>=1.5.0`**:
+   * This method always waits for both worker shutdown and inner `fnOrPromise` to resolve.
+   * If one of worker run -or- the inner promise throw an error, that error is rethrown.
+   * If both throw an error, a {@link CombinedWorkerRunError} with a `cause` attribute containing both errors.
    */
   async runUntil<R>(fnOrPromise: Promise<R> | (() => Promise<R>)): Promise<R> {
-    const runAndShutdown = async () => {
-      try {
-        if (typeof fnOrPromise === 'function') {
-          return await fnOrPromise();
-        } else {
-          return await fnOrPromise;
+    const workerRunPromise = this.run();
+    const innerPromise = typeof fnOrPromise === 'function' ? fnOrPromise() : fnOrPromise;
```

Comment:
In the function case, `fnOrPromise()` could throw an exception directly (rather than wrapped in a promise). In that case, the worker would not be shutdown.

Reply:
You can probably just reorder these two lines...

Reply:
True, if it's not an async function.
I can wrap that though, good point.